File: python/064.py
#!/usr/bin/env python3
import math

LIMIT = 10000

def period_of_sqrt_cf(n):
    a0 = int(math.isqrt(n))
    if a0 * a0 == n:
        return 0

    period = 0
    m = 0
    d = 1
    a = a0

    while True:
        m = d * a - m
        d = (n - m * m) // d
        a = (a0 + m) // d
        period += 1
        if a == 2 * a0:
            break

    return period

count = 0
for n in range(2, LIMIT + 1):
    if period_of_sqrt_cf(n) % 2 == 1:
        count += 1

print(count)

File: c/064.c
/* Project Euler 064 - Odd period square roots */
#include <stdio.h>
#include <math.h>

static int period_of_sqrt_cf(int n) {
    int a0 = (int)sqrt((double)n);
    if (a0 * a0 == n) return 0;

    int period = 0;
    int m = 0;
    int d = 1;
    int a = a0;

    while (1) {
        m = d * a - m;
        d = (n - m * m) / d;
        a = (a0 + m) / d;
        period++;
        if (a == 2 * a0) break;
    }

    return period;
}

int main(void) {
    int count = 0;

    for (int n = 2; n <= 10000; n++) {
        if (period_of_sqrt_cf(n) % 2 == 1) {
            count++;
        }
    }

    printf("%d\n", count);
    return 0;
}
